Add pep8 check to CI

Let's leave the style nitpicks to the robots.  Should be easily able to adapt xarray's here: https://github.com/pydata/xarray/blob/master/.travis.yml#L94

@micahkim23 sorry for being a little vague; I still consider #238 as a separate issue. I meant that if we go with @spencerahill's suggestion we are no longer running `flake8` on just the diff between the develop and the PR branches.  Therefore if we want the CI to only flag changes to the existing code, we need to make sure the existing code is totally clean.  

Currently it flags the following issues:
```
$ flake8 -j auto aospy
aospy/docs/conf.py:33:1: E402 module level import not at top of file
aospy/docs/conf.py:33:1: F401 'numpy' imported but unused
aospy/docs/conf.py:143:1: E402 module level import not at top of file
aospy/docs/conf.py:170:80: E501 line too long (80 > 79 characters)
aospy/aospy/automate.py:252:25: F841 local variable 'e' is assigned to but never used
aospy/aospy/automate.py:439:80: E501 line too long (96 > 79 characters)
aospy/aospy/__init__.py:2:1: F401 '.__config__.user_path' imported but unused
aospy/aospy/__init__.py:5:1: F401 '.var' imported but unused
aospy/aospy/__init__.py:7:1: F401 '.region' imported but unused
aospy/aospy/__init__.py:9:1: F401 '.run' imported but unused
aospy/aospy/__init__.py:11:1: F401 '.model' imported but unused
aospy/aospy/__init__.py:13:1: F401 '.proj' imported but unused
aospy/aospy/__init__.py:16:1: F401 '.calc.CalcInterface' imported but unused
aospy/aospy/__init__.py:16:1: F401 '.calc.Calc' imported but unused
aospy/aospy/__init__.py:17:1: F401 '.automate.submit_mult_calcs' imported but unused
aospy/aospy/__init__.py:18:1: F401 '.examples' imported but unused
aospy/aospy/model.py:153:80: E501 line too long (84 > 79 characters)
aospy/aospy/proj.py:5:1: F401 '.utils' imported but unused
aospy/aospy/calc.py:304:17: E129 visually indented line with same indent as next logical line
aospy/aospy/calc.py:686:17: E129 visually indented line with same indent as next logical line
aospy/aospy/calc.py:723:30: F821 undefined name 'PFULL_STR'
aospy/aospy/calc.py:772:1: E302 expected 2 blank lines, found 1
aospy/aospy/calc.py:783:1: E302 expected 2 blank lines, found 1
aospy/aospy/test/test_calc_basic.py:17:1: E302 expected 2 blank lines, found 1
aospy/aospy/test/test_calc_basic.py:32:1: E302 expected 2 blank lines, found 1
aospy/aospy/test/test_calc_basic.py:37:1: E302 expected 2 blank lines, found 1
aospy/aospy/test/test_calc_basic.py:129:1: E302 expected 2 blank lines, found 1
aospy/aospy/test/test_calc_basic.py:158:1: E302 expected 2 blank lines, found 1
aospy/aospy/test/test_calc_basic.py:179:17: E127 continuation line over-indented for visual indent
aospy/aospy/test/test_calc_basic.py:190:1: E305 expected 2 blank lines after class or function definition, found 1
aospy/aospy/test/__init__.py:6:9: F401 'pytest_catchlog' imported but unused
aospy/aospy/test/data/objects/examples.py:13:1: E305 expected 2 blank lines after class or function definition, found 1
aospy/aospy/utils/__init__.py:2:1: F401 '.io' imported but unused
aospy/aospy/utils/__init__.py:3:1: F401 '.times' imported but unused
aospy/aospy/utils/__init__.py:4:1: F401 '.vertcoord' imported but unused
aospy/aospy/utils/times.py:398:80: E501 line too long (107 > 79 characters)
aospy/aospy/examples/__init__.py:1:1: F401 '.example_obj_lib' imported but unused
```

Apparently some tools exist that can help automate this process, but I have not used them before:
- [autopep8](https://github.com/hhatto/autopep8)
- [autoflake](https://github.com/myint/autoflake)

--- a/.travis.yml
+++ b/.travis.yml
@@ -42,6 +42,6 @@ install:
 
 script:
   - py.test aospy --cov=aospy --cov-report term-missing --verbose
-  - flake8 -j auto xarray
+  - flake8 -j auto aospy
 after_success:
   - coveralls

--- a/.travis.yml
+++ b/.travis.yml
@@ -41,6 +41,7 @@ install:
   - pip install -e .
 
 script:
-  - py.test aospy --cov=aospy --cov-report term-missing
+  - py.test aospy --cov=aospy --cov-report term-missing --verbose
+  - git diff upstream/develop **/*py | flake8 --diff --exit-zero || true
 after_success:
   - coveralls

--- a/.travis.yml
+++ b/.travis.yml
@@ -42,6 +42,6 @@ install:
 
 script:
   - py.test aospy --cov=aospy --cov-report term-missing --verbose
-  - git diff upstream/develop **/*py | flake8 --diff --exit-zero || true
+  - flake8 -j auto xarray
 after_success:
   - coveralls

--- a/aospy/__init__.py
+++ b/aospy/__init__.py
@@ -17,5 +17,6 @@ from .calc import CalcInterface, Calc
 from .automate import submit_mult_calcs
 from . import examples
 
-__all__ = ['Proj', 'Model', 'Run', 'Var', 'Region', 'calc',
-           '_constants', 'utils']
+__all__ = ['user_path', '_constants', 'utils', 'var', 'Var', 'region',
+           'Region', 'run', 'Run', 'model', 'Model', 'proj', 'Proj',
+           'calc', 'CalcInterface', 'Calc', 'submit_mult_calcs', 'examples']

--- a/aospy/automate.py
+++ b/aospy/automate.py
@@ -249,7 +249,7 @@ def _compute_or_skip_on_error(calc, compute_kwargs):
     """
     try:
         return calc.compute(**compute_kwargs)
-    except Exception as e:
+    except Exception:
         msg = ("Skipping aospy calculation `{0}` due to error with the "
                "following traceback: \n{1}")
         logging.warn(msg.format(calc, traceback.format_exc()))

--- a/aospy/automate.py
+++ b/aospy/automate.py
@@ -436,10 +436,11 @@ def submit_mult_calcs(calc_suite_specs, exec_options=None):
               dask.distributed Client used to schedule computations.  If None
               and parallelize is True, a LocalCluster will be started.
         - write_to_tar : (default True) If True, write results of calculations
-              to .tar files, one for each :py:class:`aospy.Run` object.  These tar files have an
-              identical directory structures the standard output relative to
-              their root directory, which is specified via the `tar_direc_out`
-              argument of each Proj object's instantiation.
+              to .tar files, one for each :py:class:`aospy.Run` object.
+              These tar files have an identical directory structures the
+              standard output relative to their root directory, which is
+              specified via the `tar_direc_out` argument of each Proj
+              object's instantiation.
 
     Returns
     -------

--- a/aospy/calc.py
+++ b/aospy/calc.py
@@ -301,7 +301,7 @@ class Calc(object):
                     ds[name_int] = model_attr
                     ds = ds.set_coords(name_int)
             if (self.dtype_in_vert == 'pressure' and
-                internal_names.PLEVEL_STR in ds.coords):
+                    internal_names.PLEVEL_STR in ds.coords):
                 self.pressure = ds.level
         return ds
 
@@ -683,7 +683,7 @@ class Calc(object):
             arr = ds[region.name]
             # Use region-specific pressure values if available.
             if (self.dtype_in_vert == internal_names.ETA_STR
-                and not dtype_out_vert):
+                    and not dtype_out_vert):
                 reg_pfull_str = region.name + '_pressure'
                 arr = arr.drop([r for r in arr.coords.iterkeys()
                                 if r not in (internal_names.PFULL_STR,
@@ -720,7 +720,7 @@ class Calc(object):
                 data = np.mean(data, axis=0)[np.newaxis, :]
         if np.any(vert):
             if self.dtype_in_vert == internal_names.ETA_STR:
-                data = data[{PFULL_STR: vert}]
+                data = data[{PFULL_STR: vert}]  # flake8: noqa
             else:
                 if np.max(self.model.level) > 1e4:
                     # Convert from Pa to hPa.
@@ -769,6 +769,7 @@ class Calc(object):
             data = self.var.to_plot_units(data, dtype_vert=dtype_out_vert)
         return data
 
+
 def _add_metadata_as_attrs(data, units, description, dtype_out_vert):
     """Add metadata attributes to Dataset or DataArray"""
     if isinstance(data, xr.DataArray):
@@ -779,6 +780,7 @@ def _add_metadata_as_attrs(data, units, description, dtype_out_vert):
             _add_metadata_as_attrs_da(arr, units, description,
                                       dtype_out_vert)
         return data
+
 
 def _add_metadata_as_attrs_da(data, units, description, dtype_out_vert):
     """Add metadata attributes to DataArray"""

--- a/aospy/examples/__init__.py
+++ b/aospy/examples/__init__.py
@@ -1,1 +1,3 @@
 from . import example_obj_lib
+
+__all__ = ['example_obj_lib']

--- a/aospy/model.py
+++ b/aospy/model.py
@@ -150,8 +150,11 @@ class Model(object):
         grid_file_paths : {None, sequence of strings}, optional
             The paths to netCDF files stored on disk from which the model's
             coordinate data can be taken.
-        default_start_date, default_end_date : {None, `datetime.datetime`}, optional
-            Default start and end dates of calculations to be performed using
+        default_start_date : {None, `datetime.datetime`}, optional
+            Default start date of calculations to be performed using
+            this Model.
+        default_end_date : {None, `datetime.datetime`}, optional
+            Default end date of calculations to be performed using
             this Model.
         runs : {None, sequence of aospy.Run objects}, optional
             The child run objects of this Model

--- a/aospy/proj.py
+++ b/aospy/proj.py
@@ -2,8 +2,6 @@
 import logging
 import time
 
-from . import utils
-
 
 class Proj(object):
     """An object that describes a single project that will use aospy.

--- a/aospy/test/__init__.py
+++ b/aospy/test/__init__.py
@@ -3,7 +3,7 @@ import pytest
 
 def requires_pytest_catchlog(test):
     try:
-        import pytest_catchlog
+        import pytest_catchlog  # flake8: noqa
     except ImportError:
         return pytest.mark.skip('requires pytest_catchlog')(test)
     else:

--- a/aospy/test/data/objects/examples.py
+++ b/aospy/test/data/objects/examples.py
@@ -10,6 +10,7 @@ ROOT_PATH = os.path.dirname(__file__)
 def total_precipitation(convection_rain, condensation_rain):
     return convection_rain + condensation_rain
 
+
 precip_files = os.path.join(os.path.split(ROOT_PATH)[0], 'netcdf',
                             '000[4-6]0101.precip_monthly.nc')
 sphum_files = os.path.join(os.path.split(ROOT_PATH)[0], 'netcdf',

--- a/aospy/test/test_automate.py
+++ b/aospy/test/test_automate.py
@@ -132,7 +132,7 @@ def test_input_func_py2_py3():
         import builtins
         assert result is builtins.input
     elif sys.version.startswith('2'):
-        assert result is raw_input
+        assert result is raw_input  # flake8: noqa
 
 
 def test_user_verify():

--- a/aospy/test/test_calc_basic.py
+++ b/aospy/test/test_calc_basic.py
@@ -16,19 +16,19 @@ from .data.objects.examples import (
 
 
 def _test_output_attrs(calc, dtype_out):
-    data = xr.open_dataset(calc.path_out[dtype_out])
-    expected_units = calc.var.units
-    if calc.dtype_out_vert == 'vert_int':
-        if expected_units != '':
-            expected_units = ("(vertical integral of {0}):"
-                              " {0} m)").format(expected_units)
-        else:
-            expected_units = ("(vertical integral of quantity"
-                              " with unspecified units)")
-    expected_description = calc.var.description
-    for name, arr in data.data_vars.items():
-        assert expected_units == arr.attrs['units']
-        assert expected_description == arr.attrs['description']
+    with xr.open_dataset(calc.path_out[dtype_out]) as data:
+        expected_units = calc.var.units
+        if calc.dtype_out_vert == 'vert_int':
+            if expected_units != '':
+                expected_units = ("(vertical integral of {0}):"
+                                  " {0} m)").format(expected_units)
+            else:
+                expected_units = ("(vertical integral of quantity"
+                                  " with unspecified units)")
+        expected_description = calc.var.description
+        for name, arr in data.data_vars.items():
+            assert expected_units == arr.attrs['units']
+            assert expected_description == arr.attrs['description']
 
 
 def _test_files_and_attrs(calc, dtype_out):

--- a/aospy/test/test_calc_basic.py
+++ b/aospy/test/test_calc_basic.py
@@ -14,6 +14,7 @@ from .data.objects.examples import (
     precip, sphum, globe, sahel
 )
 
+
 def _test_output_attrs(calc, dtype_out):
     data = xr.open_dataset(calc.path_out[dtype_out])
     expected_units = calc.var.units
@@ -29,10 +30,12 @@ def _test_output_attrs(calc, dtype_out):
         assert expected_units == arr.attrs['units']
         assert expected_description == arr.attrs['description']
 
+
 def _test_files_and_attrs(calc, dtype_out):
     assert isfile(calc.path_out[dtype_out])
     assert isfile(calc.path_tar_out)
     _test_output_attrs(calc, dtype_out)
+
 
 class TestCalcBasic(unittest.TestCase):
     def setUp(self):
@@ -126,6 +129,7 @@ class TestCalcBasic(unittest.TestCase):
         calc.compute()
         _test_files_and_attrs(calc, 'reg.av')
 
+
 class TestCalcComposite(TestCalcBasic):
     def setUp(self):
         self.test_params = {
@@ -155,6 +159,7 @@ class TestCalc3D(TestCalcBasic):
             'dtype_out_vert': 'vert_int'
         }
 
+
 @pytest.mark.parametrize(
     ('units', 'description', 'dtype_out_vert', 'expected_units',
      'expected_description'),
@@ -176,7 +181,7 @@ class TestCalc3D(TestCalcBasic):
       '(vertical integral of m): m kg m^-2)', 'rain')]
 )
 def test_attrs(units, description, dtype_out_vert, expected_units,
-                expected_description):
+               expected_description):
     da = xr.DataArray(None)
     ds = xr.Dataset({'bar': 'foo', 'boo': 'baz'})
     da = _add_metadata_as_attrs(da, units, description, dtype_out_vert)
@@ -186,6 +191,7 @@ def test_attrs(units, description, dtype_out_vert, expected_units,
     for name, arr in ds.data_vars.items():
         assert expected_units == arr.attrs['units']
         assert expected_description == arr.attrs['description']
+
 
 if __name__ == '__main__':
     unittest.main()

--- a/aospy/utils/__init__.py
+++ b/aospy/utils/__init__.py
@@ -2,3 +2,5 @@
 from . import io
 from . import times
 from . import vertcoord
+
+__all__ = ['io', 'times', 'vertcoord']

--- a/aospy/utils/times.py
+++ b/aospy/utils/times.py
@@ -406,7 +406,7 @@ def ensure_time_avg_has_cf_metadata(ds):
     -------
     Dataset or DataArray
         Time average metadata attributes added if needed.
-    """
+    """  # flake8: noqa
     if TIME_WEIGHTS_STR not in ds:
         time_weights = ds[TIME_BOUNDS_STR].diff(BOUNDS_STR)
         time_weights = time_weights.rename(TIME_WEIGHTS_STR).squeeze()

--- a/ci/environment-py27-min.yml
+++ b/ci/environment-py27-min.yml
@@ -8,6 +8,7 @@ dependencies:
   - xarray
   - dask
   - distributed
+  - flake8
   - pip:
     - coveralls
     - pytest-cov

--- a/ci/environment-py27.yml
+++ b/ci/environment-py27.yml
@@ -12,6 +12,7 @@ dependencies:
   - future
   - matplotlib
   - ipython
+  - flake8
   - pip:
     - coveralls
     - pytest-cov

--- a/ci/environment-py34.yml
+++ b/ci/environment-py34.yml
@@ -13,6 +13,7 @@ dependencies:
   - future
   - matplotlib
   - ipython
+  - flake8
   - pip:
     - coveralls
     - pytest-cov

--- a/ci/environment-py35.yml
+++ b/ci/environment-py35.yml
@@ -12,6 +12,7 @@ dependencies:
   - future
   - matplotlib
   - ipython
+  - flake8
   - pip:
     - coveralls
     - pytest-cov

--- a/ci/environment-py36-xarray-dev.yml
+++ b/ci/environment-py36-xarray-dev.yml
@@ -11,6 +11,7 @@ dependencies:
   - future
   - matplotlib
   - ipython
+  - flake8
   - pip:
     - coveralls
     - pytest-cov

--- a/ci/environment-py36.yml
+++ b/ci/environment-py36.yml
@@ -12,6 +12,7 @@ dependencies:
   - future
   - matplotlib
   - ipython
+  - flake8
   - pip:
     - coveralls
     - pytest-cov

--- a/docs/whats-new.rst
+++ b/docs/whats-new.rst
@@ -56,6 +56,11 @@ Testing
 - Create Travis CI environment that tests against the xarray
   development branch. (closes :issue:`224` via :pull: `226`).
   By `Micah Kim <https://github.com/micahkim23>`_.
+- Add flake8 to Travis CI environment to check that new code
+  adheres to pep8 style. Add verbose flag to pytest test suite.
+  (closes :issue:`234` via :pull:`237`). By `Micah Kim
+  <https://github.com/micahkim23>`_.
+
 
 Dependencies
 ~~~~~~~~~~~~

